pkg/cloudlogging/puller.go

```diff
@@ -111,3 +113,34 @@ func (p *Puller) Pull(ctx context.Context, filter string, maxCount int) ([]*logg
 	}
 	return ls, nil
 }
+
+// SteamPull pulls live log entries, it won't stop until a cancel signal happens.
+func (p *Puller) StreamPull(ctx context.Context, filter string, logCh chan<- *loggingpb.LogEntry, tailLogEntriesClient loggingpb.LoggingServiceV2_TailLogEntriesClient) (chan<- *loggingpb.LogEntry, error) {
+	if err := retry.Do(ctx, p.retry, func(ctx context.Context) error {
+		req := &loggingpb.TailLogEntriesRequest{
+			ResourceNames: []string{p.resource},
+			Filter:        filter,
+		}
+		if err := tailLogEntriesClient.Send(req); err != nil {
+			return retry.RetryableError(fmt.Errorf("failed to send request: %w", err))
+		}
+
+		for {
+			resp, err := tailLogEntriesClient.Recv()
+			if errors.Is(err, io.EOF) {
+				continue
+			}
+			if err != nil {
+				return retry.RetryableError(fmt.Errorf("failed to receive response: %w", err))
+			}
+			if resp.GetEntries() != nil {
+				for _, v := range resp.GetEntries() {
+					logCh <- v
+				}
+			}
+		}
+	}); err != nil {
+		return logCh, fmt.Errorf("failed to pull log entries: %w", err)
+	}
+	return logCh, nil
```

Comment:
My previous comment was not thought through.

You don't have to return the channel. 

```go
logCh := make(chan *loggingpb.LogEntry)

// This func call will be blocking.
if err := p.StreamPull(ctx, filter, logCh); err != nil {
  // handle err
}

// Only close the channel once this func returns,
// and no more logs will be sent to the channel.
close(logCh)
```

Reply:
should we use a go routine to do the p.StreamPull?

Reply:
In test, yes. In the real command, no.